dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,429 @@
+        StringBuilder sbBlock = new StringBuilder();
+        sbBlock.append(loopOverBlockContents(bb, analysis, escapeCharacter));
+
+        // Handle case where no contents are present
+        boolean notCentered = true;
+        if (sbBlock.length() == 0) {
+            notCentered = false;
+            if (bb.getType() == Block.BlockType.SPECIAL_BLOCK) {
+                sbBlock.append(visualizeSpecialBlock((SpecialBlock) bb));
+            } else if (bb.getType() == Block.BlockType.CONDITIONAL_BLOCK) {
+                sbBlock.append(cbFooter);
+                return sbBlock.toString();
+            } else {
+                sbBlock.append(osFooter);
+                return sbBlock.toString();
+            }
+        }
+
+        // Visualize transfer input if necessary
+        if (analysis != null) {
+            // The transfer input before this block is added before the block content
+            sbBlock.insert(0, visualizeBlockTransferInput(bb, analysis));
+            if (verbose) {
+                Node lastNode = getLastNode(bb);
+                if (lastNode != null) {
+                    StringBuilder sbStore = new StringBuilder();
+                    sbStore.append(escapeCharacter).append("~~~~~~~~~").append(escapeCharacter);
+                    sbStore.append("After:");
+                    sbStore.append(visualizeStore(analysis.getResult().getStoreAfter(lastNode)));
+                    sbBlock.append(sbStore);
+                }
+            }
+        }
+        if (notCentered) {
+            sbBlock.append(escapeCharacter);
+        }
+        sbBlock.append(cbFooter);
+        return sbBlock.toString();
+    }
+
+    /**
+     * Helper method called by {@link #visualizeBlockHelper}. Iterates over the block content and
+     * visualizes all the {@link Node}s in it.
+     *
+     * @param bb The {@link Block}.
+     * @param analysis The current analysis.
+     * @param separator The separator String to use.
+     * @return The String representation of the contents of the {@link Block}.
+     */
+    protected String loopOverBlockContents(
+            Block bb, @Nullable Analysis<A, S, T> analysis, String separator) {
+
+        List<Node> contents = new ArrayList<>();
+        StringBuilder sbBlockContents = new StringBuilder();
+        boolean notFirst = false;
+
+        addBlockContent(bb, contents);
+
+        for (Node t : contents) {
+            if (notFirst) {
+                sbBlockContents.append(separator);
+            }
+            notFirst = true;
+            sbBlockContents.append(visualizeBlockNode(t, analysis));
+        }
+        return sbBlockContents.toString();
+    }
+
+    /**
+     * Helper method called by {@link #loopOverBlockContents}. If possible, add a sequence of
+     * {@link Node}s to {@code contents} for further processing.
+     *
+     * @param bb The {@link Block}.
+     * @param contents An empty list which will store {@link Node}s. TODO.
```

Comment:
Instead of requiring that the caller creates an empty list, why not create the list in this method and return the result?